resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnClusterSuite.scala

```diff
@@ -136,6 +135,39 @@ class YarnClusterSuite extends BaseYarnClusterSuite {
     checkResult(finalState, result)
   }
 
+  private def testClusterDriverBind(
+      uiEnabled: Boolean,
+      localHost: String,
+      localIp: String,
+      success: Boolean): Unit = {
+    val result = File.createTempFile("result", null, tempDir)
+    val finalState = runSpark(false, mainClassName(YarnClusterDriver.getClass),
+      appArgs = Seq(result.getAbsolutePath()),
+      extraConf = Map(
+        "spark.yarn.appMasterEnv.SPARK_LOCAL_HOSTNAME" -> localHost,
+        "spark.yarn.appMasterEnv.SPARK_LOCAL_IP" -> localIp,
+        "spark.ui.enabled" -> uiEnabled.toString
+      ))
+    if (success) {
+      checkResult(finalState, result, "success")
+    } else {
+      finalState should be (SparkAppHandle.State.FAILED)
+    }
+  }
+
+  test("yarn-cluster driver should be able to bind listeners to MM_HOST") {
```

Comment:
`NM_HOST`.

Reply:
> So I'm not sure the tests are actually that useful. The way they're written, as yarn apps, actually makes them very expensive, and this is testing basic networking config that we know will not work if the IPs are invalid.

I agree the tests are not cheap. However, they show 
- the regression of not being able to bind webUI to a specific interface is fixed 
- they demonstrate how to bind RPC and webUI to different interfaces

> The actual change you're introducing - using the bind address as the address of the NM, is actually wrong if you think about it. It just happens that the default value of that config is the local host name.

It's not wrong. it's one of the reasonable choices. Moreover it's one that is consistent with the setting executors use to bind RPC. Obviously there can be others.

> So basically if setting those env variables in the AM fixes the issue I'm not sure there's any need to change anything in Spark.

We need the fix in JettyUtils. After thinking more the change YarnRMClient.scala should use NM_HOST for registerApplicationMaster because it's the right host part of YARN's NodeId.

Reply:
> the regression of not being able to bind webUI to a specific interface is fixed

This can be a unit test for `JettyUtils` if you really care about that.

> they demonstrate how to bind RPC and webUI to different interfaces

This can be added to the docs. Nobody is going to look at unit test code to figure out how to do that.

> It's not wrong. it's one of the reasonable choices. 

It actually is, because it assumes that both the RM and the Spark app have the same configuration w.r.t. which interfaces they're binding to.

But as you say, it's better to use `$NM_HOST` to find the NM's address instead of the current code.

> We need the fix in JettyUtils

There's a separate PR with that fix too (the one I referenced above was closed and the correct one was opened as #20883).

Reply:
> it assumes that both the RM and the Spark app have the same configuration w.r.t. which interfaces they're binding to.
RM is not actively involved here. The driver executes on the NM. The launch context of the driver prescribes to assign `SPARK_LOCAL_IP` and `SPARK_LOCAL_HOSTNAME` on the worker node. Then AM rpcs the tracking URL computed on the NM back to RM.

Reply:
I will see what to do with this PR since #20883 is going faster

Reply:
> RM is not actively involved here. 

Right, I meant NM. The rest of the comment for why using Spark's bind address is not correct is still valid though.

Reply:
The URL should have a reachable authority for proxying and direct use.

Reply:
Isn't that the `trackingUrl` which is a separate parameter to the call?

Reply:
correct, SPARK_LOCAL_IP/HOSTNAME and SPARK_PUBLIC_DNS play a role in how tracking URL is generated

Reply:
Yes, but your change isn't touching that argument. I was wondering what is the effect of the code you're actually changing.

Reply:
Maybe we are talking about different parts of the patch. This thread is attached to my test code where I am demoing how this happens. It doesn't explicitly validate the generated URL's, only how bind works/fails.

Reply:
Yes, we are, sorry. I'm referring to your change in YarnRMClient.scala.